File: sorts/split-quicksort.py
#!/usr/bin/env python3
#coding=utf-8



def split(a,low,high):
	a_len=len(a)
	i=low
	x=a[low]

	for j in range(i+1,high+1):
		if a[j]<=x:
			i+=1
			if i != j:
				a[i],a[j] = a[j],a[i]

	a[low],a[i] = a[i],a[low]
	return i


def quicksort(a,low,high):
	if low < high:
		t=split(a,low,high)
		quicksort(a,low,t-1)
		quicksort(a,t+1,high)

	return a


if __name__ == '__main__':
	import random
	N=random.randint(10,40)
	A=[ random.randint(1,100) for x in range(N) ]
	print('init A[]\n',A,sep='')
	quicksort(A,0,N-1)
	print(A)

File: sorts/split-quicksort.c
#include<stdio.h>
#include<stdlib.h>

#define n 15

int A[n];

void output()
{
	int i;
	for(i=0;i<n;i++) printf("%d ",A[i]);
	puts("");
}

int split(int low,int high)
{
	int i,x,j,t;
	i=low;
	x=A[low];

	for(j=i+1;j<=high;j++)
	{
		if(A[j]<=x)
		{
			i++;
			if(i!=j)
			{
				t=A[j];A[j]=A[i];A[i]=t;
			}
		}
	}

	t=A[low];A[low]=A[i];A[i]=t;

return i;
}


int quicksort(int low,int high)
{
	int t;
	if(low<high)
	{
		t=split(low,high);
		quicksort(low,t-1);
		quicksort(t+1,high);
	}

return 0;
}


int main()
{
	int i;
	for(i=0;i<n;i++) A[i]=rand()%100;
	puts("init A[]");
	output();

	//i=split(0,n-1);
	//printf("w : %d\n",w);
	quicksort(0,n-1);

	output();
return 0;
}
